### PR TITLE
Get default params from matplotlibrc.template.

### PR DESCRIPTION
**Pull request by anntzer**
Alternate version of #14929, made a separate PR as the approach is slightly different, so that we can compare them.

This differs from #14929 in that matplotlibrc.template (and thus site-packages/matplotlib/mpl-data/matplotlibrc) remains fully commented out, but is still used as the source of defaults at runtime by using a special loader that fully uncomments it first.

-----

This makes matplotlibrc.template a fully commented-out but otherwise
valid style file representing the default style (modulo things that
styles do not control, such as the backend), avoids having to duplicate
the defaults in rcsetup, removes the need to test that the default
matplotlibrc is up to date (if not, matplotlib fails to import), and
makes it easier to fit lines within 79 characters in rcsetup.py.

The only tricky points are:
- datapath is not really an rcparam anyways, so just hack `__getitem__`
  to always return the correct value.
- the entry for path.effects was incorrect, as [] would be validated as
  the "[]" list which is not a valid list of patheffects.  In fact
  this rcParam cannot be meaningfully set from matplotlibrc; one would
  need to do an eval() like for axes.prop_cycle but let's not get
  there...  I changed the validator to validate_anylist, which at least
  works for the default of an empty list...
- we need to be a bit more careful when constructing the global rcParams
  instance as well as rcParamsOrig, rcParamsDefault, to not resolve
  _auto_backend_sentinel too early.

One can check that the rcParams are unchanged by printing them out --
that catches a typo: two entries in font.fantasy should be "Impact",
"Western"; not "ImpactWestern".

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/rcsetup.py [hunk 1]
```diff
@@ -951,484 +949,464 @@ def _validate_linestyle(ls):
...
-    'webagg.port':       [8988, validate_int],
-    'webagg.address':    ['127.0.0.1', validate_webagg_address],
-    'webagg.open_in_browser': [True, validate_bool],
-    'webagg.port_retries': [50, validate_int],
-    'toolbar':           ['toolbar2', validate_toolbar],
-    'datapath':          [None, validate_any],  # see _get_data_path_cached
-    'interactive':       [False, validate_bool],
-    'timezone':          ['UTC', validate_string],
+
+_validators = {  # key -> validator
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/rcsetup.py]: expanded the comment

**Comment by anntzer** [comment]
Quick perf check: Right now on my machine importing numpy takes 150ms, mpl+agg backend 450ms; parsing the full default matplotlibrc ~6ms.  So it's not completely negligible, but not ridiculously slow either.

**Comment by ImportanceOfBeingErnest** [comment]
If this is the route to go now, the name of `matplotlibrc.template` is pretty misleading. On the other hand, one wouldn't want to change it I guess?

File: matplotlibrc.template [hunk 2]
```diff
@@ -628,7 +628,7 @@
                      ##         line (in pixels).
                      ##     - *randomness* is the factor by which the length is
                      ##         randomly scaled.
-#path.effects : []   ##
+#path.effects :
```
Comment (ImportanceOfBeingErnest) [inline matplotlibrc.template]: Is this change necessary? If so, should one worry about it being necessary?

Comment (tacaswell) [reply matplotlibrc.template]: Indeed, the template has been wrong for a while....

Comment (anntzer) [reply matplotlibrc.template]: This previously eval'ed to the "[]" string, which is just plain wrong.  Indeed, this is one of the motivations for dogfooding our own style template.

File: lib/matplotlib/__init__.py [hunk 3]
```diff
@@ -118,6 +118,7 @@
 from collections import namedtuple
 from collections.abc import MutableMapping
 import contextlib
+from contextlib import ExitStack
```
Comment (QuLogic) [inline lib/matplotlib/__init__.py]: You already have `contextlib` imported, not sure this saves enough characters to require a second import.

File: lib/matplotlib/rcsetup.py [hunk 4]
```diff
@@ -951,484 +949,467 @@ def _validate_linestyle(ls):
...
-    'toolbar':           ['toolbar2', validate_toolbar],
-    'datapath':          [None, validate_any],  # see _get_data_path_cached
-    'interactive':       [False, validate_bool],
-    'timezone':          ['UTC', validate_string],
+
+# Mapping of rcParams to validators.  The rcParams defaults are defined in
+# matplotlibrc.template, which gets copied to matplotlib/mpl-data/matplotlibrc
+# by the setup script.
+_validators = {
+    'backend':          validate_backend,
```
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py]: I'm a bit confused over whether you want to align or not the values in this dictionary. Sometimes they are (in a block), but sometimes not.

File: lib/matplotlib/__init__.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/__init__.py]: I tend to prefer importing classes using from foo import Class, but it's not really a hard rule so I can remove this import if you prefer.

File: lib/matplotlib/rcsetup.py (hunk 4 above)
Comment (anntzer) [reply lib/matplotlib/rcsetup.py]: This was inherited from the earlier inconsistent indenting, but sure, I indented everything blockwise.

**Comment by tacaswell** [comment]
@ImportanceOfBeingErnest @QuLogic We plan to merge this tomorrow if no one protests.

**Comment by ImportanceOfBeingErnest** [comment]
I'm not protesting. But still, maybe someone has a good idea about the problem of the word *template* in the name. 

So my main point here is really this: If a user sees this as template, they may change this very file inplace. But this would have the effect that the **defaults** are changed. I.e. the sentence in the [docs](https://matplotlib.org/tutorials/introductory/customizing.html#dynamic-rc-settings) 

> The `matplotlib.rcdefaults()` command will restore the standard matplotlib default settings.

would be wrong, or is it?

**Comment by anntzer** [comment]
I clarified explicitly in https://github.com/matplotlib/matplotlib/pull/15029/files#diff-0c53b1896676a6e95e2138933242b567R5 that the file should not be edited.  Obviously people can still ignore instructions, but then they can also edit `matplotlib/__init__.py` and whatnot and we can't do anything about that.

**Comment by ImportanceOfBeingErnest** [comment]
Sure, but I guess it makes a difference if they find a file called "template" or "__init__.py". I guess the "Do NOT edit" is good enough for now.

**Comment by anntzer** [comment]
I can put it literally first line of the file if that makes you feel better :)

**Comment by ImportanceOfBeingErnest** [comment]
Naa, I mean if a file that is meant to be copied says "Do not edit this file" without any context, and this line will be copied as well.... :rofl: 
So better keep it where it is, but maybe define `it`. Which exact file(s) shall not be edited?

**Comment by anntzer** [comment]
ok, I tried to clarify that

**Comment by timhoffm** [comment]
Agreeing with @ImportanceOfBeingErnest on the naming. Something called template is *meant* to be adapted. If we want to be very explicit on the semantics we should use the template to create a defaults file. Something like https://github.com/matplotlib/matplotlib/pull/14929#issuecomment-520171523 (I’ve not been following the implementation of this PR exactly so the exact way of working may need to be adapted for it, but you’ll get the basic idea).

**Comment by anntzer** [comment]
But the file won't even be *named* matplotlibrc.template in the install.  It will be named matplotlibrc...

**Comment by ImportanceOfBeingErnest** [comment]
Yeah, I think my wording above caused some confusion. To clarify:

We have two versions of this file. The first is called `matplotlibrc.template` and it resides in the main matplotlib folder. When matplotlib is built, this file is copied to `matplotlib/lib/matplotlib/mpl-data/` and takes the name of `matplotlibrc`. Once you have built matplotlib, you *could* savely edit `matplotlibrc.template` and the confusing thing might just be that such edit have no effect. 

As I see it, the new header of the file makes this clear enough. One might also be a bit more specific about the above in the documentation though, but that can also be another PR.

**Comment by ImportanceOfBeingErnest** [comment]
Quick question: Does this PR have any effect on people who distribute vendored versions of matplotlib?

**Comment by anntzer** [comment]
In https://github.com/matplotlib/matplotlib/blob/5e9c54e7c4feb296b387fadb56394ef67ac2927f/lib/matplotlib/__init__.py#L696-L713 we already rely on a matplotlibrc file being present as last fallback, otherwise we exit with RuntimeError.  Previously they could have avoided this by putting their own matplotlibrc in the user's configdir (which sounds like a terrible idea for the vendorer), but now they can't anymore.  Really, that just means that the matplotlibrc file must be present in the install like any .py file.

**Comment by tacaswell** [comment]
pushed to 3.3

**Comment by anntzer** [comment]
I honestly do not understand why people are so worried about this file being edited.

- The previous version already warned against editing it
```
## If you edit it there, please note that it will be overwritten in your
## next install.  If you want to keep a permanent local copy that will not
## be overwritten, place it in one of the following locations:
```
- The new version explicitly tells to not edit it.  For good measure I NOW HAVE THIS IN BIG CAPITAL LETTERS at the very top.
```
#### MATPLOTLIBRC FORMAT

## NOTE FOR END USERS: DO NOT EDIT THIS FILE!
##
## This is a sample matplotlib configuration file - you can find a copy
## of it on your system in site-packages/matplotlib/mpl-data/matplotlibrc
## (which related to your Python installation location).
##
## You should find a copy of it on your system at
## site-packages/matplotlib/mpl-data/matplotlibrc (relative to your Python
## installation location).  DO NOT EDIT IT!
```
- I think if someone edits a file in the middle of their site-packages, it doesn't matter whether it's .py file or not -- it's their responsibility (thinking about it again, we could even name it matplotlibrc.py even though it's just fully commented out and keep the current parser to further inhibit it from being edited, hehe (probably a terrible idea)).
- Even if they do edit it, that "just" means that the default style is changed.  Is that really that problematic?  They could already do that by editing rcsetup.py before.

**Comment by anntzer** [comment]
rebased

**Comment by jklymak** [comment]
ping @tacaswell are you still blocking on this?  Should we discuss on a call?

File: lib/matplotlib/__init__.py [hunk 5]
```diff
@@ -855,16 +854,34 @@ def rc_params_from_file(fname, fail_on_error=False, use_default_template=True):
     return config
 
 
-# this is the instance used by the matplotlib classes
-rcParams = rc_params()
+# When constructing the global instances, we need to perform certain updates
+# by explicitly calling the superclass (dict.update, dict.items) to avoid
+# triggering resolution of _auto_backend_sentinel.
+def _load_rcs():
```
Comment (tacaswell) [inline lib/matplotlib/__init__.py]: I think this also needs a check that we have not lost any keys?

**Comment by tacaswell** [comment]
I am more convinced on this that I was over the summer.

As I said in the line comment, I think we need a check that all things that we have validators for actually exist in the dictionary.

**Comment by anntzer** [comment]
The completeness of the template is effectively being checked in the dict comprehension that builds defaultParams.  Left a comment to clarify that.

**Comment by anntzer** [comment]
@tacaswell Sorry, just realized I can make _rc_params_in_file much cleaner (and closer to the original) by adding a `transform` kwarg (a function transforming each individual line) rather than a `contents` kwarg (completely overriding the file contents), so force-pushed that.

File: lib/matplotlib/__init__.py [hunk 6]
```diff
@@ -749,19 +744,31 @@ def _open_file_or_url(fname):
             yield f
 
 
-def _rc_params_in_file(fname, fail_on_error=False):
+def _rc_params_in_file(fname, transform=lambda x: x, fail_on_error=False):
```
Comment (tacaswell) [inline lib/matplotlib/__init__.py:747]: Can you put transfrom last for API change nit-picks?

Comment (anntzer) [reply lib/matplotlib/__init__.py:747]: this is private, and logically transform comes first...

**Comment by anntzer** [comment]
rebased

**Comment by anntzer** [comment]
rebased

File: lib/matplotlib/__init__.py (hunk 6 above)
Comment (tacaswell) [reply lib/matplotlib/__init__.py:778]: fair RE it being private.
